pkg/cache/scheduler/tas_balanced_placement.go

```diff
@@ -0,0 +1,296 @@
+}
+
+func selectOptimalDomainSetToFit(s *TASFlavorSnapshot, domains []*domain, sliceCount int32, leaderCount int32, sliceSize int32, priorizeByEntropy bool) []*domain {
+	fit, optimalNumberOfDomains, _, _ := evaluateGreedyAssignment(s, domains, sliceCount, leaderCount)
+	if !fit {
+		return nil
+	}
+
+	if priorizeByEntropy {
+		sortDomainsByCapacityAndEntropy(domains)
+	}
+
+	// domain_placements[i][j][k] stores a list of domains that uses 'i' domains with
+	// ('j' leaders and 'k' workers) left to fit
+	domainPlacements := make([]map[int32]map[int32][]*domain, optimalNumberOfDomains+1)
+	for i := range domainPlacements {
+		domainPlacements[i] = make(map[int32]map[int32][]*domain)
+	}
+	domainPlacements[0][leaderCount] = map[int32][]*domain{sliceCount * sliceSize: {}}
+
+	for _, d := range domains {
+		for i := optimalNumberOfDomains; i > 0; i-- {
+			for _, beforeLeader := range slices.Sorted(maps.Keys(domainPlacements[i-1])) {
+				for _, beforeState := range slices.Sorted(maps.Keys(domainPlacements[i-1][beforeLeader])) {
+					beforePlacement := domainPlacements[i-1][beforeLeader][beforeState]
+					if beforeLeader <= 0 && beforeState <= 0 {
+						continue
+					}
+					newPlacement := make([]*domain, len(beforePlacement), len(beforePlacement)+1)
+					copy(newPlacement, beforePlacement)
+					newPlacement = append(newPlacement, d)
+					// Case 1: Pick this domain with leader
+					if beforeLeader > 0 && d.leaderState > 0 {
+						afterLeader := beforeLeader - d.leaderState
+						afterState := beforeState - d.stateWithLeader
+						if domainPlacements[i][afterLeader] == nil {
+							domainPlacements[i][afterLeader] = make(map[int32][]*domain)
+						}
+						if _, alreadyThere := domainPlacements[i][afterLeader][afterState]; !alreadyThere {
+							domainPlacements[i][afterLeader][afterState] = newPlacement
+						}
+					}
+					// Case 2: Pick this domain without leader
+					if d.sliceState > 0 {
+						afterState := beforeState - d.state
+						if domainPlacements[i][beforeLeader] == nil {
+							domainPlacements[i][beforeLeader] = make(map[int32][]*domain)
+						}
+						if _, alreadyThere := domainPlacements[i][beforeLeader][afterState]; !alreadyThere {
+							domainPlacements[i][beforeLeader][afterState] = newPlacement
+						}
+					}
+				}
+			}
+		}
+	}
+
+	bestLeaderPlacement := domainPlacements[optimalNumberOfDomains][0]
+	bestSlice := int32(-1 << 31) // minus infinity
+	var bestSlicePlacement []*domain
+
+	for _, slicesLeft := range slices.Sorted(maps.Keys(bestLeaderPlacement)) {
+		if slicesLeft > bestSlice && slicesLeft <= 0 {
+			bestSlice = slicesLeft
+			bestSlicePlacement = bestLeaderPlacement[slicesLeft]
+		}
+	}
+	return bestSlicePlacement
+}
+
+func placeSlicesOnDomainsBalanced(s *TASFlavorSnapshot, domains []*domain, sliceCount int32, leaderCount int32, sliceSize int32, threshold int32) ([]*domain, string) {
+	resultDomains := selectOptimalDomainSetToFit(s, domains, sliceCount, leaderCount, sliceSize, false)
```

Comment:
It feels like we're calling it twice in the standard flow, or once if `levelIdx == sliceLevelIdx`. Can we avoid duplication of calling this func?